File: dlx.py
'''
dlx.py

A Python implementation of Donald Knuth\'s Dancing Links algorithm.
This is a debug version with methods added for debugging.

Reference:

Donald E. Knuth: https://arxiv.org/abs/cs/0011047

'''

from typing import Any, List, Iterator, Callable, cast
COLUMN_NAME = Any
BROADCASTER = Callable[[Iterator[List[COLUMN_NAME]]], None]
MAX_OHS_COUNT = 1000

class Link(object): # pylint: disable=too-many-instance-attributes
    'doubly linked node object'
    def __init__(self) -> None:
        self.left = self
        self.right = self
        self.up = self          # pylint: disable=invalid-name
        self.down = self

    def insert_left(self, other) -> None:
        'insert other object to the left of this one'
        other.left = self.left
        other.right = self
        self.left.right = other
        self.left = other

    def insert_above(self, other) -> None:
        'insert the other object above this one'
        other.down = self
        other.up = self.up
        self.up.down = other
        self.up = other # pylint: disable=invalid-name

class Column(Link):
    'Column class'
    def __init__(self, name: COLUMN_NAME) -> None:
        Link.__init__(self)
        self.name = name
        self.size = 0

    def cover(self) -> None:
        '''
    removes column from the header list and removes all rows in the column\'s
    own list from other column lists they are in
        '''
        self.right.left = self.left
        self.left.right = self.right
        for bit0 in self.get_bits_down():
            for bit in bit0.get_other_bits_right():
                bit.down.up = bit.up
                bit.up.down = bit.down
                bit.column.size -= 1

    def uncover(self) -> None:
        'inverse of cover'
        for bit0 in self.get_bits_up():
            for bit in bit0.get_other_bits_left():
                bit.column.size += 1
                bit.up.down = bit
                bit.down.up = bit
        self.right.left = self
        self.left.right = self

    def get_bits_down(self):
        'iterate over all the bits in a column going downward'
        bit = self.down
        while bit != self:
            yield cast(Bit, bit)
            bit = bit.down

    def get_bits_up(self):
        'iterate over all the bits in a column going upward'
        bit = self.up
        while bit != self:
            yield cast(Bit, bit)
            bit = bit.up

class Bit(Link):
    'Set Bit class'
    def __init__(self, col: Column) -> None:
        Link.__init__(self)
        self.column = col

    def get_other_bits_right(self):
        'iterate over all the other bits in the row going right'
        bit = self.right
        while bit != self:
            yield cast(Bit, bit)
            bit = bit.right

    def get_other_bits_left(self):
        'iterate over all the other bits in the row going left'
        bit = self.left
        while bit != self:
            yield cast(Bit, bit)
            bit = bit.left

class Root(Link):
    'Root class -- implementing the Dancing Links algorithm'
    def __init__(self,
                 column_names: List[COLUMN_NAME],
                 set_bits: List[List[int]]) -> None:
        '''
    Argmuents:

        column_names:

            a list of objects that the caller uses as an identifier
            for the column of the dancing links matrix. When the
            function \'search\' is called the callback function is
            passed a column object and the caller can then retrieve
            the column name. The width of the dancing links matrix
            is equal to the length of the column names list

        set_bits:

            a list of lists of integers. The length of \'set_bits\'
            is equal to the number of rows in the dancing links
            matrix. Each member of set_bits is a list of occupied
            columns of the corresponding row in the dancing links
            matrix. Consider the following dancing links matrix:

                    0 1 2 3 4 5 6
                  +---------------+
                0 | 0 0 1 0 1 1 0 |
                1 | 1 0 0 1 0 0 1 |
                2 | 0 1 1 0 0 1 0 |
                3 | 1 0 0 1 0 0 0 |
                4 | 0 1 0 0 0 0 1 |
                5 | 0 0 0 1 1 0 1 |
                  +---------------+

            Since there are 6 rows in the dancing links matrix, the set_bits
            representation of this will have 6 rows, where each row
            is a list of the columns that are occupied with a bit, thus the
            set_bits represetation of this dancing links matrix is

            set_bits = [[2,4,5],[0,3,6],[1,2,5],[0,3],[1,6],[3,4,6]]
        '''
        Link.__init__(self)
        columns = []
        for column_name in column_names:
            column = Column(name=column_name)
            self.insert_left(column)
            columns.append(column)
        for row in set_bits:
            bit0 = None
            for i in row:
                column = columns[i]
                bit = Bit(column)
                column.insert_above(bit)
                column.size += 1
                if bit0 is None:
                    bit0 = bit
                else:
                    bit0.insert_left(bit)

    def _choose(self) -> Column:
        # returns the column with the least number of 1's
        smallest_size = 100000000
        ans = None
        column = self.right # type: Column
        while column != self:
            if column.size < smallest_size: # pylint: disable=no-member
                ans = column
                smallest_size = column.size # pylint: disable=no-member
            column = cast(Column, column.right)
        return ans

    def _search(self,
                k: int,
                ohs: List[Bit],
                bcaster: BROADCASTER
               ) -> None:
        # invokes the dancing links search algorithm
        if self.right == self:
            bcaster(_get_column_name_lists(ohs[:k]))
            return
        column = self._choose()
        column.cover()
        for bit_0 in column.get_bits_down():
            ohs[k] = bit_0
            for bit_1 in bit_0.get_other_bits_right():
                bit_1.column.cover()
            self._search(k + 1, ohs, bcaster)
            bit_2 = ohs[k]
            column = bit_2.column
            for bit_3 in bit_2.get_other_bits_left():
                bit_3.column.uncover()
        column.uncover()

    def search(self, bcaster: BROADCASTER) -> None:
        'searches for the solutions of the exact cover problem'
        self._search(0, [None]*MAX_OHS_COUNT, bcaster)

    def get_columns(self) -> Iterator[Column]:
        'iterate the columns of the dlx graph'
        column = self.right
        while column != self:
            yield cast(Column, column)
            column = column.right

def _get_column_name_lists(bits: List[Bit]) -> Iterator[List[COLUMN_NAME]]:
    for bit0 in bits:
        ans = [bit0.column.name]
        for bit1 in bit0.get_other_bits_right():
            ans.append(bit1.column.name)
        yield ans

File: JavascriptImplementation/dlx.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Link = /** @class */ (function () {
    // id: number;
    // static next_id : number = 0 // for debugging
    // static links : Link[] = []  // for debugging
    function Link() {
        this.left = this;
        this.right = this;
        this.up = this;
        this.down = this;
        // this.id = Link.next_id  // for debugging
        // Link.links.push(this)   // for debugging
        // Link.next_id += 1
    }
    Link.prototype.insert_left = function (other) {
        other.left = this.left;
        other.right = this;
        this.left.right = other;
        this.left = other;
    };
    Link.prototype.insert_above = function (other) {
        other.down = this;
        other.up = this.up;
        this.up.down = other;
        this.up = other;
    };
    Link.prototype.iter_down = function (f) {
        var link = this.down;
        while (link != this) {
            f(link);
            link = link.down;
        }
    };
    Link.prototype.iter_up = function (f) {
        var link = this.up;
        while (link != this) {
            f(link);
            link = link.up;
        }
    };
    Link.prototype.iter_right = function (f) {
        var link = this.right;
        while (link != this) {
            f(link);
            link = link.right;
        }
    };
    Link.prototype.iter_left = function (f) {
        var link = this.left;
        while (link != this) {
            f(link);
            link = link.left;
        }
    };
    return Link;
}());
exports.Link = Link;
var Bit = /** @class */ (function (_super) {
    __extends(Bit, _super);
    function Bit(column) {
        var _this = _super.call(this) || this;
        _this.column = column;
        return _this;
    }
    return Bit;
}(Link));
exports.Bit = Bit;
var Column = /** @class */ (function (_super) {
    __extends(Column, _super);
    function Column(index) {
        var _this = _super.call(this) || this;
        _this.index = index;
        _this.size = 0;
        return _this;
    }
    Column.prototype.cover = function () {
        this.right.left = this.left;
        this.left.right = this.right;
        this.iter_down(function (colLink) {
            colLink.iter_right(function (rowLink) {
                rowLink.down.up = rowLink.up;
                rowLink.up.down = rowLink.down;
                (rowLink.column).size -= 1;
            });
        });
    };
    Column.prototype.uncover = function () {
        this.iter_up(function (colLink) {
            colLink.iter_left(function (rowLink) {
                var bit = rowLink;
                bit.column.size += 1;
                bit.up.down = bit.down.up = bit;
            });
        });
        this.right.left = this;
        this.left.right = this;
    };
    return Column;
}(Link));
exports.Column = Column;
var Root = /** @class */ (function (_super) {
    __extends(Root, _super);
    function Root(column_count, set_bits) {
        var _this = _super.call(this) || this;
        _this.MAX_OHS_COUNT = 100000;
        var columns = [];
        for (var index = 0; index < column_count; index++) {
            var column = new Column(index);
            _this.insert_left(column);
            columns.push(column);
        }
        for (var _i = 0, set_bits_1 = set_bits; _i < set_bits_1.length; _i++) {
            var row = set_bits_1[_i];
            var bit0 = null;
            for (var _a = 0, row_1 = row; _a < row_1.length; _a++) {
                var i = row_1[_a];
                var column = columns[i];
                var bit = new Bit(column);
                column.insert_above(bit);
                column.size += 1;
                if (bit0)
                    bit0.insert_left(bit);
                else
                    bit0 = bit;
            }
        }
        return _this;
    }
    Root.prototype.choose = function () {
        var smallest_size = 1000000;
        var ans = null;
        var link = this.right;
        while (link != this) {
            var column = link;
            if (column.size < smallest_size) {
                ans = column;
                smallest_size = column.size;
            }
            link = link.right;
        }
        return ans;
    };
    Root.bcast_column_names = function (bits, bcaster) {
        var _loop_1 = function (bit0) {
            var ans = [bit0.column.index];
            bit0.iter_right(function (link) { ans.push(link.column.index); });
            bcaster(ans);
        };
        for (var _i = 0, bits_1 = bits; _i < bits_1.length; _i++) {
            var bit0 = bits_1[_i];
            _loop_1(bit0);
        }
    };
    Root.prototype.privateSearch = function (k, ohs, bcaster) {
        var _this = this;
        if (this.right == this) {
            Root.bcast_column_names(ohs.slice(0, k), bcaster);
            return;
        }
        var column = this.choose();
        column.cover();
        column.iter_down(function (link0) {
            ohs[k] = link0;
            link0.iter_right(function (link1) {
                link1.column.cover();
            });
            _this.privateSearch(k + 1, ohs, bcaster);
            var bit2 = ohs[k];
            column = bit2.column;
            bit2.iter_left(function (link2) {
                link2.column.uncover();
            });
        });
        column.uncover();
    };
    Root.prototype.search = function (bcaster) {
        var ohs = new Array(this.MAX_OHS_COUNT);
        this.privateSearch(0, ohs, bcaster);
    };
    Root.prototype.iter_columns = function (f) {
        var column = this.right;
        while (column != this) {
            f(column);
            column = column.right;
        }
    };
    return Root;
}(Link));
exports.Root = Root;
